tests/connections.py

```diff
@@ -0,0 +1,90 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the Apache 2.0 License.
+import sys
+import os
+import infra.proc
+
+import e2e_args
+import getpass
+import os
+import time
+import logging
+import multiprocessing
+from random import seed
+import infra.ccf
+import infra.proc
+import infra.jsonrpc
+import json
+import contextlib
+import resource
+import psutil
+
+from loguru import logger as LOG
+
+
+def run(args):
+    hosts = ["localhost", "localhost"]
+
+    with infra.ccf.network(
+        hosts, args.build_dir, args.debug_nodes, args.perf_nodes, pdb=args.pdb
+    ) as network:
+        primary, others = network.start_and_join(args)
+
+        primary_pid = primary.remote.remote.proc.pid
```

Comment:
could this be encapsulated in a function `primary.get_pid()` instead?

Reply:
With some extra work, because it would need to be implemented for both the LocalRemote and the SSHRemote. There's no evidence we're going to need this anywhere else, so I think it's reasonable to keep it simple for now.

Reply:
Sure, but if it's there others might know about it and use it too.

Reply:
This is an internal testing API, I’m not as worried about it as I am about our public API ;)